File: v2/file-service/internal/api/models/file.go
package models

import (
	"time"
)

// File represents a file in the system
type File struct {
	ID                string    `json:"id"`
	Filename          string    `json:"filename"`
	StoredFilename    string    `json:"storedFilename,omitempty"`
	Filesize          int64     `json:"filesize"`
	Width             int       `json:"width,omitempty"`
	Height            int       `json:"height,omitempty"`
	ThumbnailFilename string    `json:"thumbnailFilename,omitempty"`
	MimeType          string    `json:"mimeType"`
	MD5Hash           string    `json:"md5Hash"`
	SHA256Hash        string    `json:"sha256Hash,omitempty"`
	IsSpoilered       bool      `json:"isSpoilered"`
	CreatedAt         time.Time `json:"createdAt"`
	PostID            string    `json:"postId,omitempty"`
	FileURL           string    `json:"fileUrl"`
	ThumbnailURL      string    `json:"thumbnailUrl"`
}

// FileUploadResponse represents the response after a successful file upload
type FileUploadResponse struct {
	ID             string `json:"id"`
	FileURL        string `json:"fileUrl"`
	ThumbnailURL   string `json:"thumbnailUrl"`
	Filename       string `json:"filename"`
	Filesize       int64  `json:"filesize"`
	Width          int    `json:"width,omitempty"`
	Height         int    `json:"height,omitempty"`
	MimeType       string `json:"mimeType"`
	MD5Hash        string `json:"md5Hash"`
	IsSpoilered    bool   `json:"isSpoilered,omitempty"`
	UploadDuration int    `json:"uploadDuration,omitempty"`
}

// FileCheckRequest represents a request to check if a file exists
type FileCheckRequest struct {
	MD5Hash string `json:"md5Hash"`
}

// FileCheckResponse represents the response to a file check request
type FileCheckResponse struct {
	Exists bool  `json:"exists"`
	File   *File `json:"file,omitempty"`
}

// BannedHashesResponse represents the response to a banned hashes request
type BannedHashesResponse struct {
	Data      []string  `json:"data"`
	UpdatedAt time.Time `json:"updatedAt"`
}

// FileStats represents statistics about files in the system
type FileStats struct {
	TotalFiles      int            `json:"totalFiles"`
	TotalSize       int64          `json:"totalSize"`
	FilesByType     map[string]int `json:"filesByType"`
	AverageFileSize int64          `json:"averageFileSize"`
	FilesLastDay    int            `json:"filesLastDay"`
	FilesLastWeek   int            `json:"filesLastWeek"`
}

// FilePurgeRequest represents a request to purge old files
type FilePurgeRequest struct {
	OlderThanDays  int      `json:"olderThanDays"`
	MimeTypes      []string `json:"mimeTypes,omitempty"`
	ExceptBoardIds []string `json:"exceptBoardIds,omitempty"`
	DryRun         bool     `json:"dryRun,omitempty"`
}

// FilePurgeResponse represents the response to a file purge request
type FilePurgeResponse struct {
	TaskID                string `json:"taskId"`
	EstimatedFilesToPurge int    `json:"estimatedFilesToPurge"`
	EstimatedSpaceToFree  int64  `json:"estimatedSpaceToFree"`
}

File: v2/media-processor/src/models/file.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Represents a file in the system
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct File {
    /// Unique identifier for the file
    pub id: Uuid,

    /// Original filename provided by the user
    pub filename: String,

    /// Name as stored in the storage system
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stored_filename: Option<String>,

    /// Size of the file in bytes
    pub filesize: i64,

    /// Width in pixels (for images)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<i32>,

    /// Height in pixels (for images)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub height: Option<i32>,

    /// Filename of the thumbnail
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thumbnail_filename: Option<String>,

    /// MIME type of the file
    pub mime_type: String,

    /// MD5 hash of the file content
    pub md5_hash: String,

    /// SHA256 hash of the file content
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sha256_hash: Option<String>,

    /// Whether the file is marked as a spoiler
    pub is_spoilered: bool,

    /// When the file was created
    pub created_at: DateTime<Utc>,

    /// ID of the post this file is attached to
    #[serde(skip_serializing_if = "Option::is_none")]
    pub post_id: Option<Uuid>,

    /// URL to access the file
    pub file_url: String,

    /// URL to access the thumbnail
    pub thumbnail_url: String,
}

/// Response after successfully uploading a file
#[derive(Debug, Serialize, Deserialize)]
pub struct FileUploadResponse {
    /// Unique identifier for the file
    pub id: Uuid,

    /// URL to access the file
    pub file_url: String,

    /// URL to access the thumbnail
    pub thumbnail_url: String,

    /// Original filename provided by the user
    pub filename: String,

    /// Size of the file in bytes
    pub filesize: i64,

    /// Width in pixels (for images)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<i32>,

    /// Height in pixels (for images)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub height: Option<i32>,

    /// MIME type of the file
    pub mime_type: String,

    /// MD5 hash of the file content
    pub md5_hash: String,

    /// Whether the file is marked as a spoiler
    pub is_spoilered: bool,

    /// Time taken to upload in milliseconds
    #[serde(skip_serializing_if = "Option::is_none")]
    pub upload_duration: Option<i32>,
}

/// Request to check if a file exists by its MD5 hash
#[derive(Debug, Deserialize)]
pub struct FileCheckRequest {
    /// MD5 hash to check
    pub md5_hash: String,
}

/// Response to a file check request
#[derive(Debug, Serialize)]
pub struct FileCheckResponse {
    /// Whether a file with this hash exists
    pub exists: bool,

    /// The file details if it exists
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file: Option<File>,
}

/// Response containing banned file hashes
#[derive(Debug, Serialize)]
pub struct BannedHashesResponse {
    /// List of banned MD5 hashes
    pub data: Vec<String>,

    /// When the list was last updated
    pub updated_at: DateTime<Utc>,
}

/// Statistics about files in the system
#[derive(Debug, Serialize)]
pub struct FileStats {
    /// Total number of files
    pub total_files: i32,

    /// Total size of all files in bytes
    pub total_size: i64,

    /// Files grouped by MIME type
    pub files_by_type: std::collections::HashMap<String, i32>,

    /// Average file size in bytes
    pub average_file_size: i64,

    /// Number of files uploaded in the last day
    pub files_last_day: i32,

    /// Number of files uploaded in the last week
    pub files_last_week: i32,
}

/// Request to purge old files
#[derive(Debug, Deserialize)]
pub struct FilePurgeRequest {
    /// Purge files older than this many days
    pub older_than_days: i32,

    /// Optional list of MIME types to target
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_types: Option<Vec<String>>,

    /// Optional list of board IDs to exclude
    #[serde(skip_serializing_if = "Option::is_none")]
    pub except_board_ids: Option<Vec<String>>,

    /// Whether to do a dry run (no actual deletion)
    #[serde(default)]
    pub dry_run: bool,
}

/// Response to a file purge request
#[derive(Debug, Serialize)]
pub struct FilePurgeResponse {
    /// ID of the task
    pub task_id: Uuid,

    /// Estimated number of files to be purged
    pub estimated_files_to_purge: i32,

    /// Estimated space to be freed in bytes
    pub estimated_space_to_free: i64,
}
